llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp

```diff
@@ -258,13 +258,32 @@ void XtensaFrameLowering::determineCalleeSaves(MachineFunction &MF,
 
 static unsigned estimateFunctionSizeInBytes(const MachineFunction &MF,
                                             const XtensaInstrInfo &TII) {
-  unsigned FnSize = 0;
+  const Align FunctiontAlignment = MF.getAlignment();
+  /// Offset - Distance from the beginning of the function to the end
+  /// of the basic block.
+  unsigned Offset = 0;
+
   for (auto &MBB : MF) {
+    const Align Alignment = MBB.getAlignment();
+    unsigned BlockSize = 0;
+
     for (auto &MI : MBB) {
-      FnSize += TII.getInstSizeInBytes(MI);
+      BlockSize += TII.getInstSizeInBytes(MI);
     }
+
+    unsigned OffsetBB;
+    if (Alignment <= FunctiontAlignment) {
+      OffsetBB = alignTo(Offset, Alignment);
+    } else {
+      // The alignment of this MBB is larger than the function's alignment, so
+      // we can't tell whether or not it will insert nops. Assume that it will.
+      OffsetBB = alignTo(Offset, Alignment) + Alignment.value() -
+                 FunctiontAlignment.value();
```

Comment:
Is this handled somewhere else already? Should this whole size estimate function be moved to a common helper function?

Reply:
I moved this function to XtensaUtils.cpp.

Reply:
I mean outside of the backend, a generic codegen utility.

Was this edge case already handled in other code? Is there a test for it?

Reply:
I'm trying to find code which may be substituted by such function. Currently I found couple of places in backends:
https://github.com/llvm/llvm-project/blob/ca1154d1d41c75db6594428a8cdf263cf7041896/llvm/lib/Target/LoongArch/LoongArchFrameLowering.cpp#L122
and 
https://github.com/llvm/llvm-project/blob/ca1154d1d41c75db6594428a8cdf263cf7041896/llvm/lib/Target/RISCV/RISCVFrameLowering.cpp#L1216
This backends use more simple approach to caclulate function size.

And also there is a code in BranchRelaxation use similar approach to calculate block offsets, but it doesn't need function size:
https://github.com/llvm/llvm-project/blob/a518ed2d815c16010a6262edd0414a5f60a63a39/llvm/lib/CodeGen/BranchRelaxation.cpp#L68
Should we find a code in other places, which can benefit by using such function before placing it in generic utils? Or maybe leave this function it in Xtensa backend?

Reply:
This is a mess but I see RISCV is doing some target specific hacking in there I don't understand. The LoongArch one looks like a buggier version of the generic logic

Reply:
> I mean outside of the backend, a generic codegen utility.
> 
> Was this edge case already handled in other code? Is there a test for it?

I added new test for branch relaxation. May I ask your for advice about appropriate place for codegen utility? Should I place it in support or maybe add it to some class?